File: services/business-services/business-menu-option-v3/infrastructure/expose/webservice/rest/menu_option_rest_service.go
package rest

import (
	"business-menu-option-v3/application/service"
	"business-menu-option-v3/domain/model/request"
	"business-menu-option-v3/infrastructure/exception/handler"
	"github.com/gin-gonic/gin"
	"net/http"
	"strconv"
)

type MenuOptionRestService struct {
	menuOptionService service.MenuOptionService
}

func NewMenuOptionRestService(service service.MenuOptionService) *MenuOptionRestService {
	return &MenuOptionRestService{
		menuOptionService: service,
	}
}

func (restService *MenuOptionRestService) FindAll(context *gin.Context) {
	menuOptionList, err := restService.menuOptionService.FindAll()
	if err != nil {
		handler.ErrorHandler(context, err)
		return
	}
	context.JSON(http.StatusOK, menuOptionList)
}

func (restService *MenuOptionRestService) FindById(context *gin.Context) {
	id, _ := strconv.ParseUint(context.Param("id"), 10, 64)
	menuOption, err := restService.menuOptionService.FindById(uint(id))
	if err != nil {
		handler.ErrorHandler(context, err)
		return
	}
	context.JSON(http.StatusOK, menuOption)
}

func (restService *MenuOptionRestService) Save(context *gin.Context) {
	var menuOptionRequest request.MenuOptionRequest
	if err := context.ShouldBindJSON(&menuOptionRequest); err != nil {
		handler.ErrorHandler(context, err)
		return
	}

	menuOption, err := restService.menuOptionService.Save(menuOptionRequest)
	if err != nil {
		handler.ErrorHandler(context, err)
		return
	}
	context.JSON(http.StatusCreated, menuOption)
}

func (restService *MenuOptionRestService) Update(context *gin.Context) {
	id, _ := strconv.ParseUint(context.Param("id"), 10, 64)
	var menuOptionRequest request.MenuOptionRequest

	if err := context.ShouldBindJSON(&menuOptionRequest); err != nil {
		handler.ErrorHandler(context, err)
		return
	}

	menuOption, err := restService.menuOptionService.Update(menuOptionRequest, uint(id))
	if err != nil {
		handler.ErrorHandler(context, err)
		return
	}
	context.JSON(http.StatusOK, menuOption)
}

func (restService *MenuOptionRestService) Delete(context *gin.Context) {
	id, _ := strconv.ParseUint(context.Param("id"), 10, 64)
	err := restService.menuOptionService.Delete(uint(id))
	if err != nil {
		handler.ErrorHandler(context, err)
		return
	}
	context.Status(http.StatusNoContent)
}

File: services/business-services/menu-v2/src/main/java/com/demo/bbq/business/menuoption/infrastructure/resource/rest/MenuOptionRestService.java
package com.demo.bbq.business.menuoption.infrastructure.resource.rest;

import com.demo.bbq.business.menuoption.application.service.MenuOptionService;
import com.demo.bbq.business.menuoption.domain.model.request.MenuOptionSaveRequest;
import com.demo.bbq.business.menuoption.domain.model.request.MenuOptionUpdateRequest;
import com.demo.bbq.business.menuoption.domain.model.response.MenuOption;
import io.smallrye.mutiny.Uni;
import lombok.extern.slf4j.Slf4j;
import javax.inject.Inject;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.UriBuilder;
import java.net.URI;
import java.util.List;
import java.util.function.Function;

@Slf4j
@Path("/bbq/business/menu/v2/menu-options")
@Produces("application/stream+json")
@Consumes(MediaType.APPLICATION_JSON)
public class MenuOptionRestService {

  @Inject
  MenuOptionService menuOptionService;

  @GET
  @Path("/{productCode}")
  public Uni<MenuOption> findByProductCode(@PathParam("productCode") String productCode) {
    return menuOptionService.findByProductCode(productCode);
  }

  @GET
  public Uni<List<MenuOption>> findByCategory(@QueryParam("category") String categoryCode) {
    return menuOptionService.findByCategory(categoryCode);
  }

  @POST
  public Uni<Response> save(MenuOptionSaveRequest menuOptionRequest) {
    return menuOptionService.save(menuOptionRequest)
        .onItem()
        .ifNotNull()
        .transform(ignore -> Response.created(buildPostUriLocation.apply(menuOptionRequest.getProductCode())).build());
  }

  @DELETE
  @Path("/{productCode}")
  public Uni<Response> delete(@PathParam("productCode") String productCode) {
    return menuOptionService.deleteByProductCode(productCode)
        .map(isDeleted -> Response.noContent().build());
  }

  @PUT
  @Path("/{productCode}")
  public Uni<Response> update(@PathParam("productCode") String productCode, MenuOptionUpdateRequest menuOptionRequest) {
    return menuOptionService.update(menuOptionRequest, productCode)
        .onItem()
        .ifNotNull()
        .transform(ignore -> Response.created(buildPostUriLocation.apply(productCode)).build());
  }

  private final static Function<String, URI> buildPostUriLocation = productCode ->
      UriBuilder.fromResource(MenuOptionRestService.class).path("/{productCode}").build(productCode);

}
